lib/irb/pager.rb

```diff
@@ -47,12 +49,41 @@ def page(retain_content: false)
       rescue Errno::EPIPE
       end
 
-      private
-
       def should_page?
         IRB.conf[:USE_PAGER] && STDIN.tty? && (ENV.key?("TERM") && ENV["TERM"] != "dumb")
       end
 
+      def page_with_preview(width, height, modifier_proc)
+        overflow_callback = ->(lines) do
+          modified_output = modifier_proc.call(lines.join, true)
+          content, = take_first_page(width, height - 1) {|o| o.write modified_output }
+          content = content.chomp
+          content = "#{content}\e[0m" if Color.colorable?
+          $stdout.puts content
+        end
+        out = PageOverflowIO.new(width, height, overflow_callback)
+        yield out
+        content = modifier_proc.call(out.string, out.multipage?)
+        if out.multipage?
+          page(retain_content: true) do |io|
+            io.puts content
+          end
+        else
+          $stdout.puts content
+        end
+      end
+
+      def take_first_page(width, height)
```

Comment:
I hope I didn't miss anything, but can we simply take `content` as an argument instead of yielding the overlow io?

```suggestion
      def take_first_page(content, width, height)
```

I think it'll reduce the code complexity a little bit.

Reply:
Does this suggestion make sense? @tompng

Reply:
This part needs the block. Other two part that uses `take_first_page` does not need the block though.
```ruby
if omit
  content, overflow = Pager.take_first_page(winwidth, 1) do |out|
    @context.inspect_last_value(out)
  end
  ...
end
```
To show this quickly.
```
irb(main):001> a = 1000000.times.to_a
=> 
[0,
...
irb(main):002>
```
We need to take the first page in a case that full content is not available yet.